### DISCUSSION TITLE
HTTP 422 Invalid schema definition data for PROTOBUF schema

### DISCUSSION THREAD

**Discussion by raymondBourges** [apache/pulsar#19690] [Q&A] [answered]
Hi all,

I start using pulsar. I come from a project where we used kafka and avro schemas.

I try to use pulsar whith protobuf but I have a “HTTP 422 Invalid schema definition data for PROTOBUF schema” message…

My context : Pulsar-Java-v2.11.0

My schema : 
```
syntax = "proto3";

message Personne {
    string nom = 1;
    string prenom = 2;
}
```

My command (inspired by https://pulsar.apache.org/docs/2.11.x/admin-api-schemas/#upload-a-schema) : `pulsar-admin schemas upload --filename $schemaDefinitionFile public/poc-personne/personne` with `$schemaDefinitionFile` as
```
{
    "type": "PROTOBUF",
    "schema": "syntax = \"proto3\";message Personne {    string nom = 1;    string prenom = 2;}",
    "properties": {}
}
```

I try some remote debug of my pulsar running in docker and I find in the stack trace of the exception :
```
org.apache.avro.SchemaParseException: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'syntax': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (String)"syntax = "proto3";message Personne {    string nom = 1;    string prenom = 2;}"; line: 1, column: 7]
```
from `org.apache.pulsar.broker.service.schema.validator.StructSchemaDataValidator.validate()`

Could you tell me where I make a mistake ?

Thanks

**Comment by jak78** [comment, marked as answer]
Hi @raymondBourges, and welcome to the community! 
The `schema` has to be described in JSON format.
Can you try this?
```json
{
    "type":"PROTOBUF", 
    "properties":{},
    "schema":"{\"type\":\"record\",\"name\":\"Personne\",\"fields\":[{\"name\":\"nom\",\"type\":\"string\",\"default\":\"\"},{\"name\":\"prenom\",\"type\":\"string\",\"default\":\"\"}]}"
}
```

**Reply by jak78** [reply]
The human-readable version of this schema:
```json
    {
      "type": "record",
      "name": "Personne",
      "fields": [
        {
          "name": "nom",
          "type": "string",
          "default": ""
        },
        {
          "name": "prenom",
          "type": "string",
          "default": ""
        }
      ]
    }
```

**Comment by visortelle** [comment]
Hi @raymondBourges 🙂

`PROTOBUF` schema type uses AVRO schema definition format: https://docs.oracle.com/cd/E26161_02/html/GettingStartedGuide/avroschemas.html#avro-record-definitions

In order to use proto3 syntax, you can try to use `PROTOBUF_NATIVE` schema type.
https://pulsar.apache.org/docs/2.11.x/schema-get-started/#protobufnative

In my opinion, `PROTOBUF_NATIVE` schema type implementation in Pulsar has downsides in comparison to AVRO-based schema types. See https://github.com/apache/pulsar/issues/19385.

Fortunately, there is some work in progress in this direction: https://github.com/apache/pulsar/issues/19565.

**Comment by raymondBourges** [comment]
Thank you @jak78 and @visortelle,

I have a test and it's works.

But I still confusing between schema type and schema definition...

Waiting full `protobuf` support I will use `avro` for the twice.

**Reply by jak78** [reply]
@raymondBourges I'm happy you succeeded! I'll improve the doc. But you can also contribute if you want, that would be amazing :-)

**Comment by jak78** [comment]
FYI the doc has been fixed: https://pulsar.apache.org/docs/next/admin-api-schemas/
Thanks, @raymondBourges, for your feedback!
